### PR TITLE
Rename Container to Module in autogradpp and reorg code

### PR DESCRIPTION
**Pull request by goldsborough**
This PR:
1. Renames `ContainerImpl` to `Module` (`torch::nn::Module`),
2. Moves the `clone()` functionality of `Container_CRTP` into `torch::nn::CloneableModule<T>`
3. Moves the `make()` functionality of `Container_CRTP` into a free `make()` function. This is a temporary solution until we come up with a better API for module construction. I just wanted to get rid of `Container_CRTP`.
4. Removes the `AUTOGRAD_CONTAINER_CLASS` macro in favor of `: public CloneableModule<T>`
5. Reorgs code from `containers.{h,cpp}` into more fine-grained files: `rnn.{h,cpp}`, `conv.{h,cpp}` etc.
6. Removes the `Container = std::shared_ptr<Module>` in favor of the explicit type

@ebetica @jgehring

### CODE REVIEW

**Comment by ezyang** [comment]
Spurious submodule changes

File: torch/csrc/api/include/torch/nn/module.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ public:
+  // explicit CloneableModule(const char* name) : Module(name) {}
+
+  std::unique_ptr<Module> clone() const override {
+    auto ptr = std::unique_ptr<Module>(
+        new Derived(*static_cast<const Derived*>(this)));
+    ptr->children_.clear();
+    ptr->params_.clear();
+    ptr->initialize_containers();
+    ptr->initialize_parameters();
```
Comment (apaszke) [inline torch/csrc/api/include/torch/nn/module.h]: This is not a mechanism we want to keep long term, right?

**Review by apaszke — APPROVED** [review]
Are we planning to get rid of the `make()` function at some point? The current API looks really bug prone, because you can effectively have uninitialized modules, and the type system won't help you deal with that at all.

Also, what's the convention for the `/api` dir? `/api/src` are files that are only meant to be compiled, while `/api/include` are the headers we're going to install?

**Comment by goldsborough** [comment]
@apaszke Yep `make()` along with the suite of `initialize_` methods will definitely disappear step by step (trying to gradually change it so I can update tests + StarCraft). The goal is to morph these classes into what I had in #6345 essentially.

I'm putting all `.cpp` files into `src` and header files into `include`. I think in terms of distribution, we wouldn't necessarily ship these files with our Python distro, but rather folks would `submodule` PyTorch and `add_subdirectory()` it to compile it into a library, like we do for ATen.

**Review by ebetica — APPROVED** [review]
Looks okay. I'm not super happy with the verbosity and readability of some of this, but it'll do for now while you make gradual changes. I was trying to move it towards a "chain methods for everything" kind of approach, which I think is great for model readability. 

Having things simple and one template to follow that "just works" is great for research and prototyping, whereas I know it's a bit in odds with the idea of strongly typed, strong c++ style for production.

**Comment by apaszke** [comment]
Won't it be the case that if they want to use PyTorch from C++ they would need all of `torch/csrc/` and not only `torch/csrc/api/src`? Is there any reason why we want to keep includes in a separate directory than source files? That's different than all our conventions so far, and in my experience makes traversing the code much harder, because you have to jump between two deep chains of directories.

**Comment by ebetica** [comment]
Not only that, but currently the build generates some headers, so they also need another directory chain that is a copy but with generated headers.

**Comment by ebetica** [comment]
This is what our FindTorch looks like: https://gist.github.com/ebetica/2d543dbaa191c741a5ca91e423e021d0

**Comment by goldsborough** [comment]
1. Yes they would need all of `torch/csrc/`. I think it's a good idea though to keep it in its own subdirectory, just for logical separation. Ideally it would even be a top-level directory, e.g. at the level of `csrc` in order to better express that it uses but is not part of `csrc`, but I think I went for `csrc/api` originally because we have all our C++ code under `csrc` and existing build code assumes so.
2. For separate `include` and `src`: For one, autogradpp already had this structure, so I overtook. For the other, it is quite the standard C++ library organization scheme and I think it scales better to a very large number of files (which the API won't have, but still). Also, since we agreed in the old C++ API PR that don't have to care about microsecond gains from inlining, we can move towards having the headers have as little implementation as possible, serving mainly as an interface and documentation source, and move all the implementations into `.cpp` files. Then jumping between directories should actually be reduced.
3. Just to be clear, the free `make()` function is a *very* temporary solution, I'm working on a new construction design right now. However, one of the constraints will certainly be to not allow partially constructed objects, as is currently the case (pointed out by @apaszke earlier). I will post a design doc for this.

**Pull request merged**